### DISCUSSION TITLE
Smart Contract Named Actions - AEIP16

### DISCUSSION THREAD

**Discussion by samuelmanzanera** [archethic-foundation/aeip#47] [AEIP]
This is a discussion for the AEIP16 located at: https://github.com/archethic-foundation/aeip/blob/main/AEIP-16.md

**Comment by bchamagne** [comment]
I suggest an improvement on the function arguments that I believe would make them friendlier:

We could explicitly write the arguments names, ex: `on: "vote(aeip_id)"`. 
And then the `aeip_id` variable would be available in the block.

**Reply by samuelmanzanera** [reply]
I think it can be a nice idea !
The developer should then be careful to overwrite this variable, as it's not a constant if it's a variable.

**Reply by samuelmanzanera** [reply]
Is it possible to avoid this confusion @bchamagne ?

**Reply by bchamagne** [reply]
yes the `traverse` would have to keep track of the function parameters (in the acc) and check that no assignation is done in the block (by overrideing the common_interpreter on the assignation case).

**Comment by Neylix** [comment]
Looks good !
One remark, you mention 
> This field should also be present in the `condition transaction` block to be able to whitelist/blacklist the parameters.

But we don't know what is the field you're talking about.
Is think this is the `call_parameter`, but since it it in the `on` statement, you don't talk about this field.

Also, I think it's not very clear that in transaction V2 the `action` and `parameters` fields are optionnal and could be omitted to fall back in default action `actions triggered_by: transaction do ... end`

**Reply by samuelmanzanera** [reply]
I missed it, but it was thinking more about the `action` as it's mentioned in the `recipients` list from the transaction data

**Reply by samuelmanzanera** [reply]
Thing is, previous smart contract should be able to receive transaction whatever the transaction's version.
The backward compatibility should be hard requirements.
It would be strange to send transaction v1 with contract without named actions and v2 for others.

**Reply by Neylix** [reply]
Yes of course,
think we could add in the transaction constant these new fields `transaction.action` and `transaction.call_parameter`, as it they'll be available in both condition and action block.
If it's transaction V1 or V2 without these value, they'll be a nil value or empty list for parameter
